### ISSUE TITLE
Homebrew interoperability with other package managers

### ISSUE THREAD

**Issue by pcr910303** [Homebrew/brew#5397]
# A detailed description of the proposed feature

TLDR; Please make homebrew interoperable with other package managers like npm or pip.

Searching homebrew: 

```
➜  ~ brew search regex
==> Formulae
regex-opt

==> Casks
nsregextester               regexhibit

==> npm
regex                       ansi-regex
escape-string-regexp        babel-plugin-transform-es2015-sticky-regex

==> pip
auto-regex                  hachoir-regex
flake8-regex                hypothesis-regex

```

... And more.

Homebrew can manage npm, pip, and other package managers(like python pipsi, common lisp roswell, etc.
Package managers can be pluggable by a ruby script that outputs formatted output from a search query and another script that installs and logs output appropriately. Some package managers like pip, npm is available by default in vanilla homebrew and can be activated by typing `homebrew npm` or `homebrew pip` like `homebrew cask`.

Search results show all results from all activated package managers(Like above).

Installation can be done in two ways: 

1. `brew pip install jupyter`
   Installation is done by a known source(pip here), and passes the jupyter installation to pip.
2. `brew install create-react-app`
   Search through known(activated) package managers with a configurable priority, installing the first that is found. In this case, if the `brew npm` is enabled, since create-react-app isn't in brew or homebrew cask, it would be installed by npm.

Uninstallation can also be handled similarly.

# The motivation for the feature

When I(I think most homebrew users will, too) have a package to install, I usually search through homebrew first(since things like react-debugger is in homebrew cask), then (the package manager local to that language) npm, or pipsi, or pip, etc. I always thought this was natural, until I saw [this](https://raddevon.com/articles/what-are-package-managers-homebrew-and-chocolatey/) post about advantages of package managers. It tells that the advantages of package managers are that "A package manager normalizes the processes for setting up new tools", and gives an example of the AWS CLI. (It claims to be installable in both homebrew and pip). I saw this post and thought, 'Why doesn't homebrew just take care of all kinds of package manager itself?', and is writing this feature request.

# How the feature would be relevant to at least 90% of Homebrew users

Well, as I said in the motivations, I think most people that use homebrew will 'wade' through many package managers, each managing different types of packages. I think if homebrew can unify much of these managers, existing homebrew users(mostly developers, I think) will benefit by having a unificated place to query, install, and uninstall global packages. Homebrew may also accumulate users, since if this is possible, package READMEs might recommend homebrew, even packages like parcel which conventionally was installed through `npm -g`.

# What alternatives to the feature have been considered

I have no ideas about other alternatives, but some cons to consider about:
* pip and npm has packages that are intended to be an executable (which I would like to install through homebrew) but also has packages that aren't. I think the npm registry has data about if it can be executed, but I'm not sure, and showing libraries, not executables will be bad.
* While the homebrew organization has granular control about the current repo about packages, npm and pip, etc aren't controllable, and if homebrew has a policy about banned packages(maybe?) it might involve hacky solutions for homebrew to ban them.

Some other thoughts: 
* If this becomes implemented, the original homebrew cli and cask might be made an individual package manager and be plugged-in, like the other package managers, which might lead to modular code.

**Comment by MikeMcQuaid** [comment]
This isn't something that Homebrew will ever do officially I'm afraid; it would be a very large amount of work and difficult to support. It would be a good fit for an [external command](https://docs.brew.sh/External-Commands), though!

**Comment by pcr910303** [comment]
That's too bad.... I'll try to find another solution! :-(

**Comment by vitorgalvao** [comment]
> I'll try to find another solution!

Consider @MikeMcQuaid’s [suggestion](https://docs.brew.sh/External-Commands). That is exactly how `brew cask` started. It spent more years as an external command than as an official one (so far).

**Comment by pcr910303** [comment]
Okay, thaks for some insight. I’ll give it a shot. For what I understand, the external command must be written in ruby, right?
I’ll have to learn ruby first. :-(

**Comment by vitorgalvao** [comment]
> For what I understand, the external command must be written in ruby, right?

No. It can be in any language.

**Comment by pcr910303** [comment]
Ah, yes! But it’s still limited to the languages that ‘fresh’ macOS supports, right? Like python 2, or ruby, or bash/zsh, etc…

**Comment by vitorgalvao** [comment]
> But it’s still limited to the languages that ‘fresh’ macOS supports, right?

No. As long as it’s a script with a valid shebang, anything works. All this is explained [in the linked page](https://docs.brew.sh/External-Commands).

But if you’re planning on sharing it, then naturally you have to take care to be something your users have.

**Comment by pcr910303** [comment]
Ok, thanks! I’ll let you know once the POC is finished! Thanks ;-)
